homeassistant/components/miele/fan.py

```diff
@@ -0,0 +1,179 @@
+    _attr_supported_features = (
+        FanEntityFeature.SET_SPEED
+        | FanEntityFeature.TURN_ON
+        | FanEntityFeature.TURN_OFF
+    )
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return current on/off state."""
+        assert self.device.state_ventilation_step is not None
+        return self.device.state_ventilation_step > 0
+
+    @property
+    def speed_count(self) -> int:
+        """Return the number of speeds the fan supports."""
+        return int_states_in_range(SPEED_RANGE)
+
+    @property
+    def percentage(self) -> int | None:
+        """Return the current speed percentage."""
+        return ranged_value_to_percentage(
+            SPEED_RANGE,
+            (self.device.state_ventilation_step or 0),
+        )
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the speed percentage of the fan."""
+        if self.device.device_type in FAN_READ_ONLY:
+            return
+        _LOGGER.debug("Set_percentage: %s", percentage)
+        ventilation_step = math.ceil(
+            percentage_to_ranged_value(SPEED_RANGE, percentage)
+        )
+        _LOGGER.debug("Calc ventilation_step: %s", ventilation_step)
+        if ventilation_step == 0:
+            await self.async_turn_off()
+        else:
+            try:
+                await self.api.send_action(
+                    self._device_id, {VENTILATION_STEP: ventilation_step}
+                )
+            except ClientResponseError as ex:
+                raise HomeAssistantError(
+                    translation_domain=DOMAIN,
+                    translation_key="set_state_error",
+                    translation_placeholders={
+                        "entity": self.entity_id,
+                    },
+                ) from ex
+            self.device.state_ventilation_step = ventilation_step
+            self.async_write_ha_state()
+
+    async def async_turn_on(
+        self,
+        percentage: int | None = None,
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the fan."""
+        if self.device.device_type in FAN_READ_ONLY:
+            return
+        _LOGGER.debug(
+            "Turn_on -> percentage: %s, preset_mode: %s", percentage, preset_mode
+        )
+        try:
+            await self.api.send_action(self._device_id, {POWER_ON: True})
+        except ClientResponseError as ex:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="set_state_error",
+                translation_placeholders={
+                    "entity": self.entity_id,
+                },
+            ) from ex
+
+        if percentage is not None:
+            await self.async_set_percentage(percentage)
+            return
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the fan off."""
+        if self.device.device_type in FAN_READ_ONLY:
+            return
```

Comment:
What's this for?

Reply:
On the induction hub with surface integrated extraction the fan can only be monitored, not controlled (One of all quirks in the public API). This was a way in the CC to make it look the same as other fans in the UI but not generate exceptions and toast messages if touched.

Reply:
supported_features does not work in FanEntityDescription - bug or intentional? Changing to _attr_supported_features